src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ChannelFactoryBase.cs

```diff
@@ -186,5 +191,25 @@ protected override void OnEndClose(IAsyncResult result)
         {
             ChainedCloseAsyncResult.End(result);
         }
+
+        private async Task OnCloseAsyncInternal(TimeSpan timeout)
+        {
+            IChannel[] currentChannels = _channels.ToArray();
+            TimeoutHelper timeoutHelper = new TimeoutHelper(timeout);
+            foreach (IChannel channel in currentChannels)
+            {
+                IAsyncCommunicationObject iAsyncChannel = channel as IAsyncCommunicationObject;
+                if (iAsyncChannel != null)
+                {
+                    await iAsyncChannel.CloseAsync(timeout);
+                }
+                else
+                {
+                    channel.Close(timeoutHelper.RemainingTime());
+                }
+            }
+
+            await Task.Factory.FromAsync(_channels.BeginClose(timeout, null, null), _channels.EndClose, TaskCreationOptions.None);
```

Comment:
This form of FromAsync is expensive and causes an extra thread to execute in the background as it is being passed an IAsyncResult. Instead, use the FromAsync override which take a begin delegate and pass the timeout as a parameter to FromAsync.